### PR TITLE
Fix the interaction of auth and tenant modules

### PR DESCRIPTION
**Pull request by byewokko**
# Changes
- Restore the functionality of `@allow_no_tenant` decorator.
- Fix auth and tenant web wrapper order checks.
- Tenant service requires at least one tenant provider to be configured (a list of tenant ids, URL with tenant ids, or both). If there is no provider at the async stage of initialization, tenant service logs an error:
```
04-Dec-2024 14:42:50.849680 ERROR asab.web.tenant.service TenantService requires at least one provider. 
  Specify either `tenant_url` or `ids` in the [tenants] config section.
```

# Refactoring
- Tenant service config defaults moved to `asab.config`.

### CODE REVIEW

File: asab/web/tenant/utils.py [hunk 1]
```diff
@@ -37,12 +37,20 @@ def set_handler_tenant(tenant_service, route: aiohttp.web.AbstractRoute):
...
+		if tenant is None:
+			if not (hasattr(handler, "AllowNoTenant") and handler.AllowNoTenant is True):
+				raise aiohttp.web.HTTPNotFound(reason="Tenant not found.")
+			else:
+				# `None` is allowed: Tenant is optional at this endpoint.
+				pass
+
+		elif not tenant_service.is_tenant_known(tenant):
 			L.warning("Tenant not found.", struct_data={"tenant": tenant})
 			raise aiohttp.web.HTTPNotFound(reason="Tenant not found.")
```
Comment (byewokko) [inline asab/web/tenant/utils.py:55]: @ateska I think this check you introduced is not necessary here. Tenant existence has already been verified when extracting from URL path or query and setting the Tenant context variable (see the `_set_tenant_context_from_url_query` and `_set_tenant_context_from_url_path` methods below). 
The purpose of this wrapper is to simply pass the tenant context value into handler args.

File: asab/web/auth/service.py [hunk 2]
```diff
@@ -544,15 +548,17 @@ def _validate_wrapper_installation(self):
...
 				continue
 
-			if auth_wrapper_idx > tenant_wrapper_idx:
-				raise Exception("TenantService.install() must be called before AuthService.install()")
+			# Ensure the wrappers are applied in the correct order
+			tenant_wrapper_idx = tenant_service.get_web_wrapper_position(web_container)
+			if tenant_wrapper_idx is not None and auth_wrapper_idx > tenant_wrapper_idx:
+				L.error(
+					"TenantService.install(web_container) must be called before AuthService.install(web_container). "
+					"Otherwise authorization will not work properly."
```
Comment (ateska) [inline asab/web/auth/service.py:560]: This message may be misleading, b/c people are not calling `install()` explicitly, right?

File: asab/web/tenant/providers/static.py [hunk 3]
```diff
@@ -26,6 +29,7 @@ def _read_tenants_from_config(self):
 			self.Tenants.add(tenant_id)
 
 		if len(self.Tenants) > 0:
+			L.info("Static tenants loaded from config.")
```
Comment (ateska) [inline asab/web/tenant/providers/static.py:32]: debug

File: asab/web/tenant/providers/web.py [hunk 4]
```diff
@@ -46,5 +46,6 @@ async def _update_tenants(self, message_type=None):
 
 		new_tenants = set(external_tenants)
 		if self.Tenants != new_tenants:
+			L.info("Tenants from URL updated.", struct_data={"url": self.TenantUrl})
```
Comment (ateska) [inline asab/web/tenant/providers/web.py:49]: debug

File: asab/web/tenant/utils.py (hunk 1 above)
Comment (ateska) [reply asab/web/tenant/utils.py:55]: Ok, it was not working properly in SeaCat PKI

**Review by ateska — CHANGES_REQUESTED** [review]

File: asab/web/auth/service.py (hunk 2 above)
Comment (byewokko) [reply asab/web/auth/service.py:560]: this condition should only ever be true if someone calls install() explicitly.

if they rely on the automatic installation instead, this error should never occur.

**Pull request merged**